File: core/CombinationSum/CombinationSum.py
from typing import List
from copy import deepcopy


def combinationSum(candidates: List[int], target: int) -> List[List[int]]:
    res, length = [], len(candidates)

    def dfs(start, total, nums):
        if start == length:
            return
        if total == 0:
            res.append(deepcopy(nums))
            return
        dfs(start + 1, total, nums)
        if total - candidates[start] >= 0:
            nums.append(candidates[start])
            dfs(start, total - candidates[start], nums)
            nums.pop()

    dfs(0, target, [])
    return res

File: core/CombinationSum/CombinationSum.java
package com.dbc;

import java.util.ArrayList;
import java.util.List;

public class CombinationSum {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(candidates, target, res, new ArrayList<>(), 0);
        return res;
    }

    private void dfs(int[] candidates, int total, List<List<Integer>> res, List<Integer> nums, int start) {
        if (start == candidates.length) return;
        if (total == 0) {
            res.add(new ArrayList<>(nums));
            return;
        }
        dfs(candidates, total, res, nums, start + 1);
        if (total - candidates[start] >= 0) {
            nums.add(candidates[start]);
            dfs(candidates, total - candidates[start], res, nums, start);
            nums.remove(nums.size() - 1);
        }
    }
}
